File: sandbox_tests/tests/binds.c
/*
description: Mount binds are readable and not always writable
assets:
  file: ""
preexec:
  - touch /space/readonly /space/readwrite
  - bind -ro @file /space/readonly
  - bind @file /space/readwrite
*/

#include <stdio.h>
#include <stdlib.h>

int main() {
  FILE *f;

  f = fopen("/space/readonly", "rb");
  if (f == NULL) {
    perror("Failed to open /space/readonly for reading");
    return 1;
  }
  fclose(f);

  f = fopen("/space/readonly", "wb");
  if (f != NULL) {
    fprintf(stderr, "Did not fail to open /space/readonly for writing\n");
    return 1;
  }

  f = fopen("/space/readwrite", "wb");
  if (f == NULL) {
    perror("Failed to open /space/readwrite for writing");
    return 1;
  }
  fclose(f);

  return 0;
}

File: sandbox_tests/tests/binds.py
"""
description: Mount binds are readable and not always writable
assets:
  file: ""
  dir:
    file: ""
script: |
  touch("/space/readonly")
  bind_ro("file", "/space/readonly")

  touch("/space/readwrite")
  bind("file", "/space/readwrite")
  
  mkdir("/space/readonly_dir")
  bind_ro("dir", "/space/readonly_dir")
  
  mkdir("/space/readwrite_dir")
  bind("dir", "/space/readwrite_dir")

  expect(run())
"""

import os


open("/space/readonly")

try:
    open("/space/readonly", "w")
except OSError:
    pass
else:
    assert False, "Did not fail to open /space/readonly for writing"

open("/space/readwrite", "w")


os.listdir("/space/readonly_dir")

try:
    open("/space/readonly_dir/x", "w")
except OSError:
    pass
else:
    assert False, "Did not fail to open /space/readonly_dir/x for writing"

try:
    open("/space/readonly_dir/file", "w")
except OSError:
    pass
else:
    assert False, "Did not fail to open /space/readonly_dir/file for writing"

open("/space/readwrite_dir/x", "w")
open("/space/readwrite_dir/file", "w")
